bluemira/base/parameter_frame/_frame.py

```diff
@@ -115,14 +123,133 @@ def to_dict(self) -> Dict[str, Dict[str, Any]]:
     @classmethod
     def _validate_parameter_field(cls, field: str) -> Tuple[Type, ...]:
         member_type = cls.__dataclass_fields__[field].type
-        if (member_type is not NewParameter) and (
+        if (member_type is not Parameter) and (
             not hasattr(member_type, "__origin__")
-            or member_type.__origin__ is not NewParameter
+            or member_type.__origin__ is not Parameter
         ):
-            raise TypeError(f"Field '{field}' does not have type NewParameter.")
+            raise TypeError(f"Field '{field}' does not have type Parameter.")
         value_types = get_args(member_type)
         return value_types
 
+    @classmethod
+    def _validate_units(cls, param_data: Dict, value_type):
+
+        quantity = pint.Quantity(param_data["value"], param_data["unit"])
+
+        if dimensionality := quantity.units.dimensionality:
+            unit = cls._fix_combined_units(cls._remake_units(dimensionality))
+        else:
+            unit = quantity.units
+
+        unit = cls._fix_weird_units(unit, quantity.units)
+
+        if unit != quantity.units:
+            val = raw_uc(quantity.magnitude, quantity.units, unit)
+            if isinstance(param_data["value"], int) and int in value_type:
+                val = int(val)
+            param_data["value"] = val
+
+        param_data["unit"] = unit
+
+    @staticmethod
+    def _remake_units(dimensionality: Union[Dict, pint.UnitsContainer]) -> pint.Unit:
+        dim_list = list(map(base_unit_defaults.get, dimensionality.keys()))
+        dim_pow = list(dimensionality.values())
+        return pint.Unit(".".join([f"{j[0]}^{j[1]}" for j in zip(dim_list, dim_pow)]))
+
+    @staticmethod
+    def _fix_combined_units(unit: pint.Unit) -> pint.Unit:
+        dim_keys = list(combined_unit_dimensions.keys())
+        dim_val = list(combined_unit_dimensions.values())
+        with suppress(ValueError):
+            return pint.Unit(
+                combined_unit_defaults[dim_keys[dim_val.index(unit.dimensionality)]]
+            )
+        return pint.Unit(unit)
+
+    @classmethod
+    def _fix_weird_units(
+        cls, modified_unit: pint.Unit, orig_unit: pint.Unit
+    ) -> pint.Unit:
+        """
+        Essentially a crude unit parser for when we have no dimensions
+        or non-comutative dimensions
+        """
```

Comment:
This function is quite complex - I think it could probably do with some more documentation and being split up a bit.

Reply:
yeah :). If I'm honest I meant to put this as draft until i'd fixed that :sweat_smile: